trax/tf_numpy/extensions/extensions.py

```diff
@@ -453,6 +454,74 @@ def erf(x):
   return tf_np.asarray(tf.math.erf(x.data))
 
 
+# Given lhs representation, rhs representation, contraction and batch dimensions,
+# compose the output representation.
+#   e.g., ij, jk, (((1,), (0,)), ((), ())) -> ik
+#         aij, ajk, (((2,), (1,)), ((0,), (0,))) -> aik
+def compose_output_rep(lhs_rep, rhs_rep, lhs_contraction, rhs_contraction,
+                        lhs_batch, rhs_batch):
+  output_rep = []
+  for dim in lhs_batch:
+    output_rep.append(lhs_rep[dim])
+  for dim in rhs_batch:
+    if rhs_rep[dim] not in output_rep:
+      output_rep.append(rhs_rep[dim])
+
+  for i in range(len(lhs_rep)):
+    if i not in lhs_batch and i not in lhs_contraction:
+      output_rep.append(lhs_rep[i])
+  for i in range(len(rhs_rep)):
+    if i not in rhs_batch and i not in rhs_contraction:
+      output_rep.append(rhs_rep[i])
+  return ''.join(output_rep)
+
+
+# If it is the general non-batched/single-batched matrix multiplication,
+# use the highly optimized kernel `tf.tensordot` to handle it.
+def non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction):
+  return tf.tensordot(lhs, rhs, axes=(list(lhs_contraction), list(rhs_contraction)))
+
+
+# An equivalent general dot operation as that in JAX -
+#    <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dot_general.html>
+#
+# Although there is an implementation in TF XLA, avoid directly using XLA when
+# possible.
+#
+#   e.g., non-batched: ij,jk->ik
+#         batched: ijk,ikl->ijl
+def tf_dot_general(lhs, rhs, dimension_numbers):
+
+  char_list = list(string.ascii_lowercase)[8:]
+  lhs_dim, rhs_dim = len(lhs.shape), len(rhs.shape)
+  lhs_rep = char_list[:lhs_dim]
+  rhs_rep = char_list[lhs_dim:lhs_dim+rhs_dim]
+  contraction, batch = dimension_numbers
+  lhs_contraction, rhs_contraction = contraction
+  lhs_batch, rhs_batch = batch
+
+  if len(lhs_batch) == 0 and len(rhs_batch) == 0:
+    return non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction)
+
+  cond_a = lhs_dim == rhs_dim == 3
+  cond_b = lhs_batch == (0,) and rhs_batch == (0,)
+  cond_c = lhs_contraction == (lhs_dim - 1,) and rhs_contraction == (1,)
+  if cond_a and cond_b and cond_c:
```

Comment:
> I would inline cond_a/b/c because then cond_b and cond_c may not need to be evaluated because of short-circuiting.

That is a good idea! However, I tried several ways to write them inline into `if` statements but there seemed to be `SyntaxError` all the time:
```
    if ((lhs_rank == rhs_rank == 3) and (lhs_batch == (0,) and rhs_batch == (0,))) and (lhs_contraction == (2,) and rhs_contraction == (1,)):
                                                                                      ^
SyntaxError: invalid character in identifier
```
I am not sure if it is because the complex statements confuse the Python compiler.

Reply:
If I do some simple example like `if True and True and True`, then there is no such syntax error.

Reply:
It works on my machine. Which Python version are you using? And perhaps you typoed an invalid UNICODE whitespace there.